junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java

```diff
@@ -69,6 +70,27 @@
 		return testEngines;
 	}
 
+	/**
+	 * @return {@code true} if the graph does _not_ contain a cycle; else {@code false}.
+	 */
+	static boolean isAcyclic(TestDescriptor root) {
+		Set<UniqueId> visited = new HashSet<>();
+		visited.add(root.getUniqueId());
+		Stack<TestDescriptor> stack = new Stack<>();
```

Comment:
Minor nit: Could be `Deque<TestDescriptor> stack = new ArrayDeque<>` to increase performance slightly. :)

Reply:
Any link to a proof? Are we talking 10%, 1% or 0.001% here?

Reply:
> Any link to a proof?

Yep! In the Java 8 API docs for [`ArrayDeque`](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayDeque.html), the class javadocs say,

> This class is likely to be faster than Stack when used as a stack, and faster than LinkedList when used as a queue.

> Are we talking 10%, 1% or 0.001% here?

I don't know, actually! And I don't actually know how to use JMH or another benchmarking tool to show _how_ much... My bad. :P

Reply:
hehe, okay. Taking the _likely_ as solid hint. (-:

Reply:
Since you were discussing Stack vs Array, let me add my one cent. Stack is actually a subclass of Vector which is synchronized and in essence is part of the _'legacy'_ classes  preexisting the java 2 introduced collections framework. In [Vector's javadoc](https://docs.oracle.com/javase/7/docs/api/java/util/Vector.html) we can read:

> As of the Java 2 platform v1.2, this class was retrofitted to implement the List interface, making it a member of the Java Collections Framework. Unlike the new collection implementations, Vector is synchronized. If a thread-safe implementation is not needed, it is recommended to use ArrayList in place of Vector.

Also in [Stack's javadoc](https://docs.oracle.com/javase/7/docs/api/java/util/Stack.html) we can see another hint that Stack is legacy, reading:
> A more complete and consistent set of LIFO stack operations is provided by the Deque interface and its implementations, which should be used in preference to this class. 

On the other hand when we want a `stack` it is natural to use `Stack`... why would someone want to use a `Deque`..? It makes no sense... :flushed:

Reply:
@gaganis I know, it's very weird and the complete opposite of intuitive. This is one of the many drawbacks of Java's policy on maintaining backwards compatibility as much as possible.

I don't know why they originally implemented Stack by extending Vector all those years ago, or why they made it a concrete class instead of an interface which would have allowed alternative stack implementations. However, as I think you've noticed, due to those historical decisions and how Stack consequently uses `synchronized` internally, Stack does perform not as well as ArrayDeque as a stack in single-thread situations.

As for why we're only provided with a `Deque` interface and not also a super-interface that provides just `push()` and `pop()`, I've no idea as to why it was done that way. Seems a bit silly to me, personally.

...Not sure what else to say. :)